alerting/src/main/kotlin/org/opensearch/alerting/transport/TransportDocLevelMonitorFanOutAction.kt

```diff
@@ -0,0 +1,1186 @@
+                            "Monitor ${request.monitor.id} : " +
+                                "Doc Level query ${it.id} : ${it.query} doesn't have queryFieldNames populated. " +
+                                "Cannot optimize monitor to fetch only query-relevant fields. " +
+                                "Querying entire doc source."
+                        )
+                        break
+                    }
+                    fieldsToBeQueried.addAll(it.queryFieldNames)
+                }
+                if (fieldsToBeQueried.isNotEmpty()) {
+                    log.debug(
+                        "Monitor ${monitor.id} Querying only fields " +
+                            "${fieldsToBeQueried.joinToString()} instead of entire _source of documents"
+                    )
+                }
+            }
+
+            fetchShardDataAndMaybeExecutePercolateQueries(
+                monitor,
+                indexExecutionContext,
+                monitorMetadata,
+                inputRunResults,
+                docsToQueries,
+                updatedIndexNames,
+                concreteIndicesSeenSoFar,
+                ArrayList(fieldsToBeQueried),
+                shardIds.map { it.id }
+            ) { shard, maxSeqNo -> // function passed to update last run context with new max sequence number
+                indexExecutionContext.updatedLastRunContext[shard] = maxSeqNo
+            }
+            if (transformedDocs.isNotEmpty()) {
+                performPercolateQueryAndResetCounters(
+                    monitor,
+                    monitorMetadata,
+                    updatedIndexNames,
+                    concreteIndicesSeenSoFar,
+                    inputRunResults,
+                    docsToQueries,
+                )
+            }
+            monitorResult = monitorResult.copy(inputResults = InputRunResults(listOf(inputRunResults)))
+
+            /*
+             populate the map queryToDocIds with pairs of <DocLevelQuery object from queries in monitor metadata &
+             list of matched docId from inputRunResults>
+             this fixes the issue of passing id, name, tags fields of DocLevelQuery object correctly to TriggerExpressionParser
+             */
+            queries.forEach {
+                if (inputRunResults.containsKey(it.id)) {
+                    queryToDocIds[it] = inputRunResults[it.id]!!
+                }
+            }
+
+            val idQueryMap: Map<String, DocLevelQuery> = queries.associateBy { it.id }
+
+            val triggerResults = mutableMapOf<String, DocumentLevelTriggerRunResult>()
+            // If there are no triggers defined, we still want to generate findings
+            if (monitor.triggers.isEmpty()) {
+                if (dryrun == false && monitor.id != Monitor.NO_ID) {
+                    createFindings(monitor, docsToQueries, idQueryMap, true)
+                }
+            } else {
+                monitor.triggers.forEach {
```

Comment:
Prolly a gap in my understanding but can you clarify the following:
if monitor has triggers especially if it has multiple triggers.. do we generate findings multiple times? shouldn't we be executing createFindings() before "runForEachDocTrigger" right?

Reply:
this may need logic change but not a new issue introduced by this pr.